design-proposals/klip-47-add-source-stream-table-semantic.md

```diff
@@ -0,0 +1,122 @@
+when a user creates a table the DB creates the corresponding files on disk and owns the table. 
+User may write data into those tables via `INSERT INTO` statements and/or modify/delete existing data 
+via `UPDATE` and `DELETE` statements. However, the table will be not changed by any other means. 
+
+For ksqlDB, the setup is quite differently. Specifically, user could create the database based on 
+the topics stored in Kafka as the source of truth. The defined database could be mutable by 
+using `INSERT` command to populate data into it, which will then pollute the original data source.
+Secondly, when calling `CREATE TABLE`, the actual data is neither materialized nor 
+ready for interactive queries to access. To address these gaps, we would like to add:
+
+* Syntax to add read-only stream/table as `CREATE SOURCE STREAM/TABLE` to make the data source as read-only. Any 
+insertion command will be rejected, with an example error response as:
+
+```
+Error: insertion into source stream/table is not allowed
+```   
+
+* In addition, the `CREATE SOURCE TABLE` will materialize the input source topic into a stream 
+instance which would generate a RocksDB instance for interactive queries.
+
+
+## What is in scope
+* Add `CREATE SOURCE STREAM/TABLE` syntax to KSQL
+* `DESCRIBE ... EXTENDED` should show source streams and tables as read-only  
+* Add materialization of the `source table` to make it pull-query accessible
+* Make source stream/table immutable from KSQL by rejecting insertion requests
+ 
+
+## What is not in scope
+* We will not add role-based write permission for source stream/table in v1, even though it seems to be useful.
+
+## Design
+
+### Semantic Change
+We shall first add optional keyword `SOURCE` into the KSQL codebase, by adding the new syntax to `sqlBase.g4` for:
+* createStream
+* createStreamAs
+* createTable
+* createTableAs
+
+Then inside `AstBuilder#visitCreateStream` and `AstBuilder#visitCreateTable`, the built classes for stream/table 
+instances would be `CreateStream` and `CreateTable`, which are the only two subclasses of `CreateSource`. It makes sense 
+to add the `SOURCE` as a flag into `CreateSource`, and let `AstBuilder` pass the flag into the struct 
+for later stage reference of read-only.
+
+In addition, when user calls `DESCRIBE ... EXTENDED`, the read-only attribute shall be displayed for stream/table. 
+
+### Make Source Stream/Table Read-only
+We plan to add restriction check into the `DistributingExecutor#throwIfInsertOnReadOnlyTopic` or create a similar function, 
+which could verify whether the given stream/table is read-only. If so, the insertion will be rejected.
+
+### Source Table Materialization
+In the call `EngineExecutor#execute`, we will check whether given plan has a `queryPlan` before deciding to proceed 
+creating a persistent query. To bypass this check, we need to provide a concrete query plan within the input plan. We would 
```

Comment:
Does creating a persistent query mean we're creating not just a RocksDB for the materialization but also a new sink topic? If so, what will the sink topic be named? How will we protect against users deleting the new sink topic out of confusion, and causing the query to fail as a result?

Reply:
There should not be a sink topic. I guess the term "persistent query" is tricky -- we basically need to run a Kafka Streams program and "persistent query == Kafka Streams" program, even if this Kafka Streams program does not really run a _query_... Guess we don't have a good term for this special case.

Reply:
+1 for Matthias, we don't plan to add a sink topic.

Reply:
we might also want to be explicit here about changelog topics - I suspect we _do_ want to create one here?

Reply:
I am not 100% sure if we would need a changelog topic... If we assume that the input data is partitioned correctly (we did discuss the issue if users specify a "partial key"...) we could restore from the input topic. Of course, we would require that the input topic is configured with log compaction enabled.

Again, it seems we need to figure out what assumption about "valid" vs "invalid" patterns we want to apply/enforce to make a decision on this question.

Also note, if people use the "source table" in a query, we won't be able to share the state store anyway atm, but each query would read the input topic by itself, so we can design the `CST` "program" independent (for now). Of course, we also need to make sure that pull queries work correctly.

Reply:
I'm worried about not having a changelog here for the same reason as https://github.com/confluentinc/ksql/issues/5673